**Проект SQL**

  **Цель**
    
    Цель исследования — проанализировать базу данных крупного сервиса для чтения книг по подписке. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

 **Задания**

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;




  **Описание данных**

**Таблица books**

Содержит данные о книгах:

    book_id — идентификатор книги;
    author_id — идентификатор автора;
    title — название книги;
    num_pages — количество страниц;
    publication_date — дата публикации книги;
    publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

    author_id — идентификатор автора;
    author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:

    publisher_id — идентификатор издательства;
    publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

    rating_id — идентификатор оценки;
    book_id — идентификатор книги;
    username — имя пользователя, оставившего оценку;
    rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:

    review_id — идентификатор обзора;
    book_id — идентификатор книги;
    username — имя пользователя, написавшего обзор;
    text — текст обзора.



In [33]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [34]:

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [35]:
query = ''' SELECT * 
            FROM books '''

books = pd.io.sql.read_sql(query, con = engine)
display(books.head(5))
display(books.info())


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

In [36]:
query = ''' SELECT * 
            FROM authors '''

authors = pd.io.sql.read_sql(query, con = engine)
display(authors.head(5))
display(authors.info())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

In [37]:
query = ''' SELECT * 
            FROM publishers '''

publishers = pd.io.sql.read_sql(query, con = engine)
display(publishers.head(5))
display(publishers.info())

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

In [38]:
query = ''' SELECT * 
            FROM ratings '''

ratings = pd.io.sql.read_sql(query, con = engine)

display(ratings.head(5))
display(ratings.info())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

In [39]:
query = ''' SELECT * 
            FROM reviews '''

reviews = pd.io.sql.read_sql(query, con = engine)

display(reviews.head(5))
display(reviews.info())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

    Таблица books: Содержит данные о книгах
    Таблица authors: Содержит данные об авторе
    Таблица publishers: Содержит данные об издательствах
    Таблица ratings: Содержит данные о пользовательских оценках книг
    Таблица reviews: Содержит данные  о пользовательских обзорах на книги

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [40]:
query = ''' SELECT count(book_id) 
            FROM books
            WHERE CAST(publication_date AS date) >= '2000-01-01'
            '''

pd.io.sql.read_sql(query, con = engine)

count
0    821

После 1 января вышло 821 шт книг

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [41]:
#Посчитаем среднее количество обзоров
query = ''' SELECT book_id,
                   ROUND(AVG(rating))
                   
                   
            FROM ratings
            GROUP BY book_id
            ORDER BY book_id;
            
            '''

rev_avg = pd.io.sql.read_sql(query, con = engine)
display(rev_avg.head(5))
#Посчитаем среднюю оценку
query = ''' SELECT book_id,
                   --ROUND(AVG(rat.rating))
                   COUNT(review_id) as count_review
                   
            FROM reviews
            GROUP BY book_id
            ORDER BY book_id;
            
            '''

rat_count = pd.io.sql.read_sql(query, con = engine)
display(rat_count.head(5))


book_id  round
0        1    4.0
1        2    3.0
2        3    5.0
3        4    5.0
4        5    4.0

book_id  count_review
0        1             2
1        2             1
2        3             3
3        4             2
4        5             4

In [42]:
query = ''' SELECT books.book_id as book,
                   AVG(ratings.rating) AS avg_rating,
                   COUNT(DISTINCT reviews.review_id) as count_review
            FROM books
            LEFT JOIN ratings ON ratings.book_id = books.book_id
            LEFT JOIN reviews ON reviews.book_id=books.book_id
            GROUP BY books.book_id
            ORDER BY count_review DESC;
        '''

general = pd.io.sql.read_sql(query, con = engine)
display(general.head(5)) 
                   

book  avg_rating  count_review
0   948    3.662500             7
1   963    3.977273             6
2   734    4.206897             6
3   302    4.414634             6
4   695    4.081081             6

Посчитали среднюю оценку для книги и количество обзоров.

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [43]:
query = ''' SELECT publisher
            FROM (
            SELECT pub.publisher,
                   COUNT(publisher) AS number_pub
            FROM books AS b
            JOIN publishers AS pub ON pub.publisher_id=b.publisher_id
            WHERE num_pages >= 50
            GROUP BY publisher
            ORDER BY number_pub DESC
            ) AS tab
            LIMIT 1;
            
            '''

publish_count = pd.io.sql.read_sql(query, con = engine)
display(publish_count.head(5))


publisher
0  Penguin Books

Penguin Books - издательство, в котором наибольшее количество толстых книг

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [44]:
query = '''SELECT authors.author AS author_name,
                  ROUND(AVG(tab.average_rating), 2) AS avg_rate    
           FROM books
           LEFT JOIN (
                   SELECT book_id,
                          COUNT(rating) AS cnt,
                          AVG(rating) AS average_rating
                   FROM ratings
                   GROUP BY book_id
                   ) AS tab
                ON tab.book_id = books.book_id
           LEFT JOIN authors ON authors.author_id = books.author_id
           WHERE tab.cnt >= 50
           GROUP BY author_name
           ORDER BY avg_rate DESC
           LIMIT 1;   
        '''
author_top = pd.io.sql.read_sql(query, con = engine)
display(author_top.head(5))

author_name  avg_rate
0  J.K. Rowling/Mary GrandPré      4.28

Joahn Rowling имеет самую высокую среднюю оценку, учитывая только книги с 50 и более оценками.

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**

In [45]:
query = """SELECT ROUND(AVG(tab_2.review_cnt)) AS average_number_of_reviews
           FROM (
                SELECT username  
                FROM ratings
                GROUP BY username
                HAVING COUNT(rating_id) > 50) AS tab
           LEFT JOIN (
                       SELECT username,
                              COUNT(review_id) AS review_cnt
                       FROM reviews
                       GROUP BY username) AS tab_2
            ON tab_2.username = tab.username
        """
user_review_top = pd.io.sql.read_sql(query, con = engine)
display(user_review_top.head(5))

average_number_of_reviews
0                       24.0

24 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Общий вывод**

  - Penguin Books - издательство, в котором наибольшее количество толстых книг
  - Joahn Rowling имеет самую высокую среднюю оценку, учитывая только книги с 50 и более оценками
  - 24 - среднее количество обзоров от пользователей, которые поставили больше 50 оценок.
  
    Возможно, стоит предусмотреть систему поощерений для пользователей, которые ставят оценки и пишут отзывы.